components/script/dom/htmloptionscollection.rs

```diff
@@ -0,0 +1,194 @@
+            }
+        } else {
+            // Step 1
+            self.Remove(index as i32)
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-htmloptionscollection-length
+    fn Length(&self) -> u32 {
+        self.collection.Length()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-htmloptionscollection-length
+    fn SetLength(&self, length: u32) {
+        let current_length = self.collection.Length();
+        let delta = length as i32 - current_length as i32;
+        if delta < 0 {
+            // new length is lower - deleting last option elements
+            for index in (length..current_length).rev() {
+                self.Remove(index as i32).unwrap()
+            }
+        } else if delta > 0 {
+            // new length is higher - adding new option elements
+            self.add_new_elements(delta as u32).unwrap();
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-htmloptionscollection-add
+    fn Add(&self, element: HTMLOptionElementOrHTMLOptGroupElement, before: Option<HTMLElementOrLong >) -> ErrorResult {
+        // TODO
+        let root = self.root.get();
+
+        let node = match element {
+            HTMLOptionElementOrHTMLOptGroupElement::HTMLOptionElement(ref element) => element.upcast::<Node>(),
+            HTMLOptionElementOrHTMLOptGroupElement::HTMLOptGroupElement(ref element) => element.upcast::<Node>(),
+        };
+
+        // Step 1
+        if node.is_ancestor_of(root.r()) {
+            return Err(Error::HierarchyRequest);
+        }
+
+        if let Some(HTMLElementOrLong::HTMLElement(ref before_element)) = before {
+            // Step 2
+            let before_node = before_element.upcast::<Node>();
+            if !root.r().is_ancestor_of(before_node) {
+                return Err(Error::NotFound);
+            }
+
+            // Step 3
+            if node == before_node {
+                return Ok(());
+            }
+        }
+
+        // Step 4
+        let reference_node = before.and_then(|before| {
+            match before {
+                HTMLElementOrLong::HTMLElement(element) => Some(Root::upcast::<Node>(element)),
+                HTMLElementOrLong::Long(index) => {
+                    self.collection.IndexedGetter(index as u32).map(Root::upcast::<Node>)
+                }
+            }
+        });
+
+        // Step 5
+        let parent = if let Some(reference_node) = reference_node.r() {
+            reference_node.GetParentNode().unwrap()
+        } else {
+            root
+        };
+
+        // Step 6
+        Node::pre_insert(node, parent.r(), reference_node.r()).map(|_| ())
+    }
+
+    // https://html.spec.whatwg.org/multipage/#dom-htmloptionscollection-remove
+    fn Remove(&self, index: i32) -> ErrorResult {
+        if let Some(element) = self.collection.IndexedGetter(index as u32) {
+            element.r().Remove();
+            Ok(())
+        } else {
+            Err(Error::NotFound)
```

Comment:
The spec doesn't say that you should throw a NotFoundError in this case, instead it says to abort the algorithm.